skins/cat17/address-form/src/store.ts

```diff
@@ -0,0 +1,138 @@
+                isValid: Validity.INCOMPLETE
+            },
+            street: {
+                dataEntered: false,
+                value: '',
+                isValid: Validity.INCOMPLETE
+            },
+            postcode: {
+                dataEntered: false,
+                value: '',
+                isValid: Validity.INCOMPLETE
+            },
+            city: {
+                dataEntered: false,
+                value: '',
+                isValid: Validity.INCOMPLETE
+            },
+            country: {
+                dataEntered: false,
+                value: '',
+                isValid: Validity.INCOMPLETE
+            }
+        }  
+	},
+	mutations: {
+        [STORE_VALUES](state, payload): void {
+            payload.forEach((field: FormData) => {
+               state.form[field.name].value = field.value;
+           });
+        },
+        [VALIDATE_INPUT](state, payload): void {
+            payload.forEach((field: FormData) => {
+                if (field.value === '' && field.optionalField === true) {
+                    state.form[field.name] = {
+                        ...state.form[field.name],
+                        dataEntered: false,
+                        isValid: Validity.INCOMPLETE
+                    };
+                }
+                else {
+                    state.form[field.name] = {
+                        ...state.form[field.name],
+                        dataEntered: true,
+                        isValid: Helper.inputIsValid(field.value, field.pattern)
+                    };
+                }
+            });
+		},
+		[MARK_EMPTY_FIELD_INVALID](state, payload): void {
+			payload.forEach((field: FormData) => {
+                if (!field.optionalField && state.form[field.name].isValid === Validity.INCOMPLETE) {
+                    state.form[field.name].isValid = Validity.INVALID;
+				}
+			});
+		},
+		[BEGIN_ADDRESS_VALIDATION](state): void {
+			state.isValidating = true;
+		},
+        [FINISH_ADDRESS_VALIDATION](state, payload: ValidationResult): void {
+			ADDRESS_FIELD_VALIDATOR_NAMES.forEach(name => {
+                if (state.form[name].dataEntered === true) {
+                    state.form[name] = {
+                        ...state.form[name],
+                        isValid: state.form[name].isValid !== false && !payload.messages[name] ? Validity.VALID : Validity.INVALID
+					};
+				}
+			});
+			state.isValidating = false;
+		}
+	},
+	getters: {
+		validity(state, contentName: string) {
+            return state.form[contentName].isValid;
+		}
+	},
+	actions: {
+        storeAddressFields({ commit }, payload: Payload) {
+            commit('STORE_VALUES', payload.formData);
+            commit('VALIDATE_INPUT', payload.formData);
+            commit('MARK_EMPTY_FIELD_INVALID', payload.formData);
+            if (Object.keys(this.state.form).filter(field => this.state.form[field].isValid === Validity.INVALID ).length === 0 ) {
```

Comment:
The purpose if this line of code is to start the validation when all the required fields are filled, right? I think the logic is more complex than what's here and should go into a getter called `allAddressFieldsAreFilled`.

That getter should have an internal list of required fields for person/company address type and return true/false if all the required field contain a non-empty `value`. The `isValid` property is not relevant here IMHO.

Also, this part of the code is prone to concurrency errors at the moment: when multiple `storeAddress` actions come in, multiple `BEGIN_ADDRESS_VALIDATION` and `END_ADDRESS_VALIDATION` will be triggered and might resolve in the wrong order. There are several ways on how to solve this:
* Check `isValidating` property and cancel all running requests
* Keep track of the running requests with an internal counter (incremented and decremented by `BEGIN_ADDRESS_VALIDATION` and `END_ADDRESS_VALIDATION`) and make `isValidating` a getter.